### ISSUE TITLE
More cleanup.

### ISSUE THREAD

**Issue by jdickey** [jdickey/new_poc#240]
Following on from #162 as per [this comment](https://github.com/jdickey/new_poc/issues/162#issuecomment-77384196), we have two remaining cleanup activities after [giving up on CodeClimate improvement](https://github.com/jdickey/new_poc/issues/162#issuecomment-77621178):

1. Factor out common code as called out in the [Issue #54 comment](https://github.com/jdickey/new_poc/issues/54#issuecomment-77142902);
1. Spend a day or two looking at the specs and code to find *sensible* DRYing up/opportunities for custom matchers and/or support methods (as per #169).

Then, as noted earlier, we can go on to Issues #211, #169 and #203. *That* would close all currently-open issues not labelled either `fixlater` or `wontfix` *except for* #54 — which is pending on #232, about which I have *no clue at all*.

**Comment by jdickey** [comment]
This issue touches on [this week-old comment](https://github.com/jdickey/new_poc/issues/54#issuecomment-77354969) on Issue #54, and substantially overlaps Issue #243. Ick.

**Comment by jdickey** [comment]
Well, #243 isn't a consideration any more; see the comment that I closed it with.

**Comment by jdickey** [comment]
What remaining cleanup is needed after closing out #252?

#### Finish with Entities

We still have session entities remaining from the old way of doing things; consistency applies there, too. Fortunately, they're not as complicated as either our present `User` or `Post` entities.

#### Drag back `lib/newpoc-services-markdown_html_converter` into `services`

Currently ~50 lines of actual code, half of which is canned setup for the [`html-pipeline` Gem](https://github.com/jch/html-pipeline) we use. *Very* straightforward.

At this point, the only things left in `lib` are `action_support` (a grab bag of support classes shared across `Action` usage) and things like `current_user_identity.rb`, which could each use a new home but aren't presently worth stopping and holding everything until I think of one. (If anyone else has defensible suggestions, please feel free to chime in.)

Also at this point, we're completely free of the "unbundled Rails dependency" failed experiment, with a more self-consistent/understandable architecture. This also, as a *huge* bonus, allows us to dispense with the jerry-rigged `bin/rakeall.sh` script that is used for CI to run specs against the main app and all the unbuilt pseudo-Gems formerly living in `lib/newpoc-whatever`.

#### DRY up Actions

Since Actions were initially implemented as unbundled pseudo-Gems (see above), there's a lot of duplicate code that would just cause maintenance headaches (read: "defects") going forward. Fix those.

And finally...

#### Specs

We've now got a useable architecture with code that is almost entirely decoupled from direct Rails dependencies. So why are we still booting up the full Rails stack for *each* example in *each* spec file for our *entire test suite*? Feature specs, whose *raison d^ètre* is to "run the app the way a user would", of course need the full stack. Controller specs need enough of the stack to get their head on straight, but since they (should) only exercise Actions and respond to their results, those can be run less regularly (as with the feature specs). Running the remaining specs currently takes anywhere from 15 to 40 seconds to execute 342 examples. (By contrast, 252 controller specs take 10-12 seconds, and it takes 20-25 seconds to run 69 feature specs). That first number is *several* times slower than what could be achieved were Rails' complete stack not reloaded and initialised for each of those examples.

This situation has arisen because we've historically largely treated the auto-generated `spec/spec_helper.rb` file as a black box (a privilege accorded to *no other spec or code file in the main app*), whereby all sorts of slow code (database transactions around every example? code-coverage recording for every example? reporting to CodeClimate, etc?) is slowing us down. Means of addressing this that have been widely documented and discussed include:

1. Splitting out a separate `spec_helper` file, often called `spec_helper_lite` or `unit_spec_helper`, for use by specs not directly dependent on Rails (everything except feature specs, and possibly controllers) when not run as part of full integration/CI testing;
1. Use of explicit paths for `require` statements and eliminating pervasive autoloading (except in known slow specs, which are dragged down by other factors). This would also serve to explicitly document what code was needed by what other code;
1. Using RSpec's `tag` feature to mark slow specs as not to be executed by default (and modifying our CI setup to enable those specs). This should require modifying a single line in each affected spec.

**Comment by jdickey** [comment]
The last two comments  copied in full from PR #254 (which replaced #252, if you'd lost track):

----

Here's a question: *How pervasively is `Newpoc::Entity::Post` tied into the rest of the app?*

### Named and...

1. `app/controllers/posts_controller/action/create.rb`
1. `app/controllers/posts_controller/create_failure_setup.rb`
1. `app/controllers/posts_controller/error_message_builder.rb`
1. `app/controllers/posts_controller.rb`
1. `app/entity_factories/post_factory.rb`
1. `spec/controllers/posts_controller/action/create_spec.rb`
1. `spec/controllers/posts_controller_spec.rb`
1. `spec/entity_factories/post_factory_spec.rb`
1. `spec/helpers/posts_helper_spec.rb`
1. `spec/repositories/post_repository_spec.rb`

### Explained and...

Those are all the files that `grep` says include the string `Newpoc::Entity::Post`, but how exactly are they used?

1. `app/controllers/posts_controller/action/create.rb` specifies it as the default for the `PostController::Action::Create#initialize` method's `entity_class` parameter, which is used in a direct call to `entity_class.new` in the `#validate_post_data` method. That *should* be using the `PostFactory.create` method instead;
1. `app/controllers/posts_controller/create_failure_setup.rb` has the exact same usage as the above, with its direct call to `entity_class.new` in its `#rebuild_entity` method;
1. `app/controllers/posts_controller/error_message_builder.rb` directly creates a new `Newpoc::Entity::Post` in the callback built by the `.bad_attributes_matcher` class method. Also should be using the `PostFactory`;
1. `app/controllers/posts_controller.rb` defines `Newpoc::Entity::Post` as the `entity_class` to be passed as a parameter to the `Action::New` (and therefore eventually `Action::Create`) class constructors. Factory would help again;
1. `app/entity_factories/post_factory.rb` contains the *only* class that should specify the entity class directly. It *is* the factory;
1. `spec/controllers/posts_controller/action/create_spec.rb` checks that a successfully-created post *is* a `Newpoc::Entity::Post`; it also recreates an entity (via `Newpoc::Entity::Post.new`) and interrogates its validity and errors. Use of the `PostFactory` class and of ordinary duck typing would be useful here;
1. `spec/controllers/posts_controller_spec.rb` tests in several places that a purported entity is in fact an instance of `Newpoc::Entity::Post`, rather than of the class created by `PostFactory`.
1. `spec/entity_factories/post_factory_spec.rb` invokes an RSpec shared example using the line `it_behaves_like 'an entity factory for', Newpoc::Entity::Post`. *This is arguably the only spec that should explicitly name the Post class*;
1. `spec/helpers/posts_helper_spec.rb` creates entity instances by directly invoking `Newpoc::Entity::Post.new` similarly to the spec for `Action::Create`. It, too, should use the `PostFactory`; and, finally
1. `spec/repositories/post_repository_spec.rb` defines an `entity` assigned variable as an instance of `Newpoc::Entity::Post`. It, too, should use the `PostFactory`.

### Planned and...

Given all this, the way forward seems fairly clear:

1. Update the non-factory code with direct knowledge of the entity class and its construction to instead reference the `PostFactory.create` and `.entity_class` methods;
1. Once that gives us a green bar, change `PostFactory.entity_class` to return `Entity::Post` rather than `Newpoc::Entity::Post` and catalogue the carnage;
1. Order that catalogue into a dependency tree ("*A* broke because *B* broke because...").
1. Switch the entity class back and ensure that we still have a green bar;
1. **Either:**
  1. Work from the root of that dependency tree to generalise knowledge of the entity and to identify features not yet implemented in the new entity *and associated code*. On each iteration, ensure that the changes work with the previous entity definition before iterating; **or*
  1. Work from the leaf nodes of the dependency tree, identifying features not yet implemented in the new entry *and associated code*; implement and iterate, ensuring that a green bar is preserved when using the old entity class
1. After ensuring that all existing code and specs pass using the new entity class, the old `lib/newpoc-entity-post` tree can safely be removed;
1. Once *that* is done, the original `post-entities` branch can be deleted and PR #252 closed (but *not* merged).

### Questioned

Now, should we merge `post-entities-again` into `master` and re-branch, or continue on our (very) long-in-the-tooth PR that's (arguably) to a point comparable to the abandoned [first attempt](https://github.com/jdickey/new_poc/tree/post-entities/)?

----
----

And, after further deliberation, we've decided to merge and then proceed with *separate* PRs and branches for the steps under **Planned and...*, above, for

1. Step 1, updating the non-factory code;
1. Steps 2-4, building the entity dependency tree;
1. Steps 5-7; iterating that tree, removing the unbuilt Gem and the dead branch/PR.

We *may* choose to put limits on the duration of the PR/branch for that last step, merging and rebranching with a green bar after, say, 7 calendar days or ~30 commits, always with a green bar. There's too many moving parts still under the hood for us to be out in the wilderness for forty years (or forty commits), as our recent RuboCop experiences reminded us.

----

So we'll be riffing through a series of *shorter*, limited PRs to finish up Entities as above (earlier in *this* issue). *Finally*.

**Comment by jdickey** [comment]
PR #257 has been merged, completing Step 1 of our 7-step plan above. Now, what happens when the "truth" of "what is the entity class named" changes when nothing else does?

**Comment by jdickey** [comment]
...and the current answer to that question is "146 of 767 examples, in 24 of 85 spec files in the main app break", or roughly 19% of examples in 28% of files. The affected spec files are:

1. controllers/posts_controller/action/create_spec.rb
1. controllers/posts_controller_spec.rb
1. entity_factories/post_factory_spec.rb
1. features/attempting_to_update_post_invalid_data_fails_spec.rb
1. features/draft_post_only_accessible_to_author_spec.rb
1. features/member_can_create_draft_post_directly_navigate_spec.rb
1. features/member_can_create_draft_post_spec.rb
1. features/member_can_create_draft_post_view_in_profile_spec.rb
1. features/member_can_create_valid_image_post_spec.rb
1. features/member_can_create_valid_text_post_spec.rb
1. features/member_can_enter_image_url_on_new_post_form_spec.rb
1. features/member_can_publish_article_view_on_landing_page_spec.rb
1. features/member_can_publish_articles_view_on_profile_spec.rb
1. features/member_can_publish_then_edit_article_spec.rb
1. features/member_cannot_create_post_without_title_spec.rb
1. features/new_post_form_has_correct_button_text_spec.rb
1. features/new_post_form_status_defaults_draft_spec.rb
1. features/only_author_can_view_draft_post_in_profile_spec.rb
1. features/post_only_editable_by_author_spec.rb
1. features/user_can_view_article_from_landing_page_spec.rb
1. features/user_can_view_author_profile_from_article_page_spec.rb
1. helpers/posts_helper_spec.rb
1. helpers/users_helper_spec.rb
1. repositories/post_repository_spec.rb

Things to notice here:

1. 18 of the 24 files listed are feature specs (of 37 total feature specs); those will break if *anything* else does, and can be set aside for now (though we need to remember to alternate `Entity::Post` CI builds (that break) with `Newpoc::Entity::Post` CI builds (that should *always* pass));
1. The post factory spec broke (dealing with pulling attributes out and pushing them back in);
1. The create-post action spec was the only action spec that broke;
1. The `PostsHelper` spec broke, unsurprisingly;
1. The `UsersHelper` spec broke, when building a user's article list tried to call `#to_hash` on a `PostDao`; and
1. The `PostRepository' spec broke, also when calling `#to_hash` on a `PostDao`.

Again, priorities:

1. Figure out a clean way to fix the `PostFactory` specs; without those, nothing else is going to work;
1. Fix `PostRepository`;
1. Fix the `PostsController` `Create` action and the controller itself;
1. `PostsHelper`;
1. `UsersHelper`; and, finally
1. If the feature specs don't Just Work after everything else does, figure out why not and fix it (without breaking the earlier fixes).

That should keep me busy for a few days.

**Comment by jdickey** [comment]
You know, if I wasn't swimming through my sinuses with this persistent Singapore flu again, or if I had a [pair](http://c2.com/cgi/wiki?PairProgramming), we'd have shipped this by the end of last week. Taking weekends off would likely help, too.

Anybody following along and paying attention should have noticed that what led us to (prematurely, it now appears) abandon the earlier `post-entities` branch (lack of separation between presentation (Rails) details and application (entities+actions) logic) was what we were heading towards again here as soon as we started adding presentational details (`#pubdate_str` and code more explicitly packaged as "decorations"; direct use of ActiveModel, etc). Once again, we're now at the point where everything "works" (as in giving a green bar on unit test suites) but *doesn't* work from the browser (as convincingly demonstrated by a 580-line error report running `bundle exec rspec spec/features` against the latest commit when using the new `Entity::Post` entity).

At some point before it hits the presentation layer (Rails views/helpers), these entities we're passing around need to be translated back to implementation objects (i.e., DAOs). In times long past, we had an explicit translation layer set up to do just that. Where should that go in our current architecture? In the action classes, as a post-processing step? Or in the logic within the Rails controller methods like `PostsController#on_create_success` that actually set instance variables on the controller to be picked up by the views/helpers (a `$DEITY`-we-hope-it's-unique-to-Rails idiom)?

When I phrase the last paragraph as I did, I *hope* it's obvious to the reader what I think now. I'm willing to be convinced I'm not thinking through something well enough, but I see actions, like entities, as part of the (Rails-independent) application, and controllers, being *part of* Rails, as the implementation detail. We *should* be able to take our actions, entities and entity factories from, say, Commit 7ad3d2, plop them down into a RubyMotion app or a Grape-implemented JSON API or an insert-framework-*here* app, reimplement the repository layer to the existing API and the rest of a pretty plain-vanilla app around that concoction, and have everything Just Work *without touching anything that was brought over*; that brought-over code *is* the "application". (I wouldn't even try to do copy-and-paste code, either; that application logic *should* be in (real, traditionally-built) *Gems*. That takes the application code *completely out of* the deliverables for that new app, and explicitly makes it what it really is — something the delivery app depends on.)

So we're looking at what we've been calling our controller-action "responder" methods as the place for that entity-to-DAO switch to happen. And *that* brings us back to another smell we've been noticing in our code since we adopted the current architecture. Those responder methods have been getting longer and more complex, which we've "handled" by creating ad-hoc objects like [`CreateFailureSetup`](https://github.com/jdickey/new_poc/blob/7ad3d24/app/controllers/posts_controller/create_failure_setup.rb) and [`ErrorMessageBuilder`](https://github.com/jdickey/new_poc/blob/7ad3d24/app/controllers/posts_controller/error_message_builder.rb), implying via [namespace modules](http://rubymonk.com/learning/books/1-ruby-primer/chapters/35-modules/lessons/80-modules-as-namespaces) that these supporting classes are used exclusively by, if not "part of", the [`PostsController`](https://github.com/jdickey/new_poc/blob/7ad3d24/app/controllers/posts_controller.rb) itself. There's got to be a better way.

My current thinking is that reifying that "action responder" idiom is a good candidate for that better way. That would allow us to rework our [`PostsController#on_create_failure`](https://github.com/jdickey/new_poc/blob/7ad3d24/app/controllers/posts_controller.rb#L70-L75) method, for example, from

```ruby
  def on_create_failure(payload)
    @post = CreateFailureSetup.new(payload).cleanup.entity
    render 'new'
  rescue RuntimeError => e # not logged in as a registered user
    redirect_to root_path, flash: { alert: e.message }
  end
```

to

```ruby
  def on_create_failure(payload)
    Responder::CreateFailure.new(self).respond_to(payload)
  end
```

"What," you splutter. "Where's the `render` call? The redirection? What does this thing *do*?" It responds to the reported result of the application action in an implementation-appropriate manner; if you want more detail than that, read the source (or the doc, once we get staffed up sufficiently that the team stops cutting corners by not writing doc). "OK," you wonder, "but how *does* it, as a PORO, *do* the redirection or the rendering?" That's why the controller instance is passed in to the constructor (`Responder::CreateFailure.new(self)`); it leverages the [controller](http://api.rubyonrails.org/classes/ActionController/Base.html) (which leverages its own internal helpers, etc). It only needs to know how to tell *the controller* what to do. This is the (hopefully) final missing piece of our application/implementation that we've been dancing around for *two whole months now*. :anguished: 

Lesson (re)learned, example &#35;728 for the project. Blast and damn.

**Comment by jdickey** [comment]
This has dragged on *far* too long; cleanup still remains(!), but it can be done (and described!) in a far more granular manner branching off `master`. With the merging of PR #258, we now have separation between the application logic (actions+entities) and implementation (responders+DAOs) of the app using a delivery mechanism (Rails).

## Woot!!! :clap:
